source/common/router/router.h

```diff
@@ -67,6 +67,41 @@ class FilterUtility {
     bool hedge_on_per_try_timeout_;
   };
 
+  class StrictHeaderChecker {
+  public:
+    struct HeaderCheckResult {
+      bool valid_ = true;
+      const Http::HeaderEntry* entry_;
+    };
+
+    static const HeaderCheckResult test(Http::HeaderMap& headers,
+                                        const Http::LowerCaseString& target_header);
+
+    using ParseRetryFlagsFunc = std::function<uint32_t(absl::string_view)>;
+
+  private:
+    static HeaderCheckResult hasValidRetryFields(Http::HeaderEntry* header_entry,
+                                                 const ParseRetryFlagsFunc& parseFn) {
+      HeaderCheckResult r;
+      if (header_entry) {
+        const auto flags = parseFn(header_entry->value().getStringView());
```

Comment:
Thinking about this some more, maybe it would make more sense to have the parseFn return `std::pair<uint32_t, bool>` (or something to that effect, potentially a named struct), containing both the retry flags and whether there were any unknown fields? That way we avoid having to clear the bit in the other case and the interface might be a bit self explanatory?

Sorry for the churn here, hopefully we can get this to a place where we're all happy with it

Reply:
No worries re: churn. I think a named struct would be clearer at calling sites, but going to repush with `std::pair` and see how it reads during review. I agree that the clearing of the `UNKNOWN` flag is an implementation detail callers shouldn't need to remember.